rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,179 @@
+# 000 - RFC Template
+
+## Current Status
+
+### Proposed
+
+2020-03-24
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/alantech/alan/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Colton Donnelly <colton@alantechnologies.com>
+
+## Summary
+
+Rework the front end of the compiler to utilize the Hindley-Milner type system for type resolution, as well as generating Microstatements only once the program has been "solved".
+Users will be able to opt-in to the new compiler front end while it's being developed by changing their Alan file to the `.lnn` extension.
+
+## Expected SemBDD Impact
+
+If Alan was 1.0.0, this entire change could be implemented as a single patch update, since the proposed changes only affect the internal architectural changes in the compiler.
+However, the intention is to gradually move towards the new compiler front-end, which will be implemented over the course of multiple patch updates.
+As such, this change will impact many patch updates, until the point it's able to replace the current front-end without any regressions to user code.
+
+Until the new front-end is able to replace the current one, bdd tests will gradually be included in a second run which will compile programs with the new front-end.
+
+## Proposal
+
+This RFC proposes a new compiler front-end, with the most significant difference being the use of a Hindley-Milner-based type resolution system.
+By modifying the Hindley-Milner inference systems to use Alan's interface and generic types, the compiler will be able to determine the types of functions without type annotations, while also allowing the compiler to select one of many functions depending on the input types being passed in.
+To explain how the new front-end will differ from the current one, we'll use examples that grow in complexity.
+
+It's worth noting that, since this PR is talking about the inner architecture of the compiler front-end, there is terminology that is used that is specific to the compiler's current architectural design.
+Some quick definitions:
+- Microstatement: The atom of the current `lntoamm` phase's internal language. The set of Microstatements is a superset of the `amm` language - a valid `amm` program is a valid program of Microstatements, but not all Microstatements are valid in `amm`.
+
+Also, "constraint" is used in literature related to Hindley-Milner type checking to signify that a narrowing-down on the type of a value, expression, or function.
+A type can only be selected if it matches constraints, and if there are constraints that conflict with each other, that results in an invalid program.
+
+The initial motivating example is as follows:
+```ln
+from @std/app import print, start, exit
+
+on start {
+  const myMaybe: Maybe<int64> = none();
```

Comment:
Correct me if I'm wrong, but this particular example would not compile without the type hint by the user, correct?

Reply:
That's correct, I only used `none` instead of `some` to avoid walking through argument passing since I wanted a simpler example first.